src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.DataFlowOperationVisitor.cs

```diff
@@ -0,0 +1,131 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidMultipleEnumerations.FlowAnalysis;
+
+namespace Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidMultipleEnumerations
+{
+    public partial class AvoidMultipleEnumerations
+    {
+        internal abstract class AvoidMultipleEnumerationsFlowStateDictionaryFlowOperationVisitor : GlobalFlowStateDictionaryFlowOperationVisitor
+        {
+            private readonly WellKnownSymbolsInfo _wellKnownSymbolsInfo;
+
+            protected AvoidMultipleEnumerationsFlowStateDictionaryFlowOperationVisitor(
+                GlobalFlowStateDictionaryAnalysisContext analysisContext,
+                WellKnownSymbolsInfo wellKnownSymbolsInfo) : base(analysisContext)
+            {
+                _wellKnownSymbolsInfo = wellKnownSymbolsInfo;
+            }
+
+            protected abstract bool IsExpressionOfForEachStatement(SyntaxNode node);
+
+            public override GlobalFlowStateDictionaryAnalysisValue VisitParameterReference(IParameterReferenceOperation operation, object? argument)
+            {
+                var value = base.VisitParameterReference(operation, argument);
+                if (!IsDeferredType(operation.Parameter.Type.OriginalDefinition, _wellKnownSymbolsInfo.AdditionalDeferredTypes))
+                {
+                    return value;
+                }
+
+                return CreateAndUpdateAnalysisValue(operation, value);
```

Comment:
Nit: It seems like even this can be extracted into a helper if you pass along `IOperation` and underlying `ISymbol` for the operation.